design/authenticators/authn_oidc/oidc_configuration_enhancements.md

```diff
@@ -0,0 +1,250 @@
+endpoint:
+  - validate authenticator status and provider connectivity
+  - fetch provider keys for decoding tokens
+
+This should not interfere with the proposed solution, but it is important to note since the [modified discovery code](https://github.com/cyberark/conjur/blob/48e95904a5ee8cda1503db9f5744ff6eefcecbdb/app/domain/authentication/o_auth/discover_identity_provider.rb#L31C40-L31C40) will be 
+used across different authenticators to invoke the provider discovery endpoint `/.well-known/openid-configuration`. 
+It will be necessary that the other authenticators are unaffected by this change.
+
+### Security
+The solution proposes adding a CA certificate (or chain) to the OpenSSL truststore of the Conjur container for the duration of certain HTTP requests. This involves writing certificate contents to the container volume, creating a symlink in the OpenSSL truststore, and yielding for the duration of the specific logic. Once this logic is complete, the CA certificate and symlink will be removed.
+
+#### Security Risks and Mitigations
+**Certificate Management**: Temporary changes to truststores can lead to mismanagement of certificates, resulting in leftover or unused certificates remaining in the truststore. Also we need to be sure that there are no circumstances in which the certificate management process is interrupted.
+
+**Mitigation**: Implement a thorough certificate management test suite that validates the addition, usage, and removal of certificates.
+
+**Log Verbosity and Information Disclosure**: Debugging TLS issues with an OIDC provider is overly difficult with the current debug logging. We would like to expose additional logging from the OpenIDConnect gem (which makes requests to the provider) and OpenSSL (which performs TLS verification based on the configured truststore). This requires that we do not disclose any sensitive information with the additional debug logs.
+
+**Mitigation**: Verify that logs for each of the components do not disclose sensitive data, including certificate keys, metadata, or filepaths.
+
+### Testing
+The testing will depend on the final implementation, but at the very least the wrapper method will need to be 
+well-covered by unit tests.
+
+#### Unit tests (temporary truststore modifications)
+1. OpenSSL truststore contains the CA cert symlink while yielding
+1. OpenSSL truststore symlink is removed after yielding
+1. Hash collision in truststore does not produce an error but instead increments the file extension of the symlink
+1. Cert chains work in addition to individual CA certs
+1. Malformed certificates produce the expected error message
+1. Temp file/dir/symlink exists while yielding
+1. Temp file/dir/symlink contains expected certificate content while yielding
+1. Temp file/dir/symlink is removed after yielding
+
+#### Unit tests (authenticator data object)
+1. Default `ca-cert` value is nil
+1. Set value matches expected value
+
+#### Cucumber tests
```

Comment:
To do this we would need an Identity or Okta instance using a private CA which seems like a large ask from infra. Our Keycloak server uses already uses a self-signed cert which is functionally the same as a private CA anyway.

I don't know that we necessarily need an automated test for the environment variables in this approach, since we'd be testing whether a standard Ruby library (Net::HTTP) handles them correctly as opposed to anything within Conjur. Similarly for the configurable `ca-cert` change, OpenSSL doesn't care whether a trusted cert belongs to a CA or a proxy server so we'd be double-testing in a sense. 

If we do need tests against a proxy server we would likely have to setup an nginx server with a self-signed cert as a forward proxy for Conjur

Reply:
@gl-johnson do we currently have any e2e or integration tests for following:
> self-signed cert (keycloak or okra/identity)
> certificate chain (keycloak or okra/identity)
> https_proxy & HTTPS_PROXY
> concurrent writing of requests to cert store (are we making any backup of this file beforehand?  Is there any risk of file corruption and are we leveraging built in linux tools - update-ca-certs/trust)

Reply:
Keycloak is already configured with a self-signed cert which will be easy to test against. For a cert chain we'd need to generate a CA/intermediate chain and reload Keycloak with an updated [server config](https://github.com/cyberark/conjur/blob/1c84fa10e7b3b6a8d754d1a3cce7875f5d5819ad/ci/oauth/keycloak/standalone.xml#L41) - although I think it is sufficient cover cert chains in unit tests. We can easily test that a cert chain is parsed correctly and that the cert store contains all certs in the chain.

Support for HTTPS_PROXY/https_proxy is not a standard, nor is it a function of Conjur - rather the underlying library that makes the outgoing connection. The previous library chose to use the same proxy (`http_proxy`) for both HTTP and HTTPS requests which I imagine is the case 99% of the time. I personally think it is overkill to have e2e tests requiring an additional service when we can easily confirm via some manual smoke testing.

[Ruby's Tempfile library](https://docs.ruby-lang.org/en/2.4.0/Tempfile.html) ensures uniqueness so concurrent read/write operations should not be possible as far as the cert itself is concerned. We will need to add some logic to ensure we handle collisions of the symlinks properly should they happen, but if anything they would link back to the same cert (possibly stored in multiple locations) and still allow the connection to succeed.

Reply:
Thanks @gl-johnson for response 

I'm fine with #1 (keycloak) so long as we do individual cert and chain which I believe we are already doing/planning!! 👍   

For #2 https_proxy seems like it would be good candidate for a happy path e2e test - I understand this is a 3rd party library but would be good to have test that verifies this all works and if gem changes we will be notified of any breakage in pipeline vs manual testing from release to release.  let me know what you think..

For # 3 seems like this is handled already - do we have any tests to confirm there is no issue?

For K8s Follower - need to understand if this is valid scenario and how a rootless user would update certstore if connecting to K8s-follower for authn-oidc v2

Reply:
If it was something under our control I would agree, but ultimately the proxy determination in this case is handled by a standard library in Ruby. [This code is extremely unlikely to change.](https://ruby-doc.org/stdlib-3.0.2/libdoc/uri/rdoc/URI/Generic.html#method-i-find_proxy), plus I imagine we already make this assumption with every other network call that Conjur makes to an external provider. It works like so:
1. Construct env var name using scheme of the request, should result in either `http_proxy` or `https_proxy`
2. If HTTP, favor `http_proxy` if it exists, otherwise use `HTTP_PROXY`
3. If HTTPS, favor `https_proxy` if it exists, otherwise use `HTTPS_PROXY`

The older version of OpenIDConnect used [the not very well-maintained httpclient](https://github.com/nahi/httpclient/blob/4658227a46f7caa633ef8036f073bbd1f0a955a2/lib/httpclient.rb#L1071), which loaded the proxy settings itself and does not consider the scheme of the request when looking at environment variables.

As far as I can tell the K8s follower does not handle authn-oidc requests, but if you have information otherwise then I can look into it further.